Reduce datapath from_lxc complexity

jrajahalme opened the pull request:
Split per packet load balancing into a separate tail call. This reduces datapath complexity when per packet load balancing is needed.

Preparatory commits simplify tail call logic and refactors IPv6 from LXC code paths to be more similar to IPv4. This makes edits and reviews on the functional change commits easier.

This PR should have minimal effect on the code when per packet load balancing is not required. On IPv4 there should be no change, and on IPv6 there is one extra data revalidation due to logic being split into two tail callable functions.

This is a prerequisite for #18894

jrajahalme committed:
datapath: Reduce from LXC complexity

Split from LXC code paths into two tail calls when per packet load
balancing is needed. When per packet load balancing is not needed this
should have minimal impact on datapath performance.

Kernel 4.9 verifier was unhappy when service load balancing code was
removed from handle_ipv6_from_lxc() and replaced with simpler state
restoration call, but only if host firewall was enabled, and per-packet
LB was on. Had to shuffle code around to make verifier happy again.

Shuffled IPv4 code to keep them similar, but git diff looks scarier
there.

Finally had to conditionally apply a revalidate data call to make
verifier happy also in the verifier complexity tests (test-verifier).

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
datapath: Fix indent

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
datapath: Fix policy_mark_skip comment

Setting the policy skip mark will not skip deny policies. Fix the
comment to state that.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
datapath: Structure IPv6 from LXC path more like IPv4

Sharing similar code structure makes comparing code easier.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
datapath: Remove space after cast from bpf_lxc

Checkpatch complains about spaces after casts, remove them to reduce
checkpatch failures.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
datapath: Simplify lxc tail calls

Datapath has always IPv6 or IPv4 enabled, no need to check for
that. Simplifying this logic makes the code easier to read and
understand.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

kkourt commented on bpf/bpf_lxc.c:
```diff
@@ -1075,7 +1061,7 @@ ipv6_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label, __u8 *reason,
 				  0, ifindex, 0, monitor);
 		if (tuple_out)
-			memcpy(tuple_out, &tuple, sizeof(tuple));
+			*tuple_out = tuple;
```
Could it be that the difference here was intentional due to the different sizes between v4 and v6 tuples?

pchaigno replied:
I think this is indeed intentional to rely on [our own `memcpy` variant](https://github.com/cilium/cilium/blob/4f27865e975cc095891b68df5f4a7bc0e26ecca5/bpf/include/bpf/builtins.h#L154), which can optimize the number of instructions necessary for the copy.

jrajahalme replied:
OK, reverted back.

kkourt commented on bpf/bpf_lxc.c:
```diff
@@ -470,8 +470,7 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
 
-declare_tailcall_if(__or3(is_defined(ENABLE_IPV4), is_defined(ENABLE_IPV6),
-			  is_defined(DEBUG)), CILIUM_CALL_IPV6_FROM_LXC)
+__section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_IPV6_FROM_LXC)
```
FWIW, I believe the original intention of this was to maintain the code structure of the previous version when we checked for an OR instead of an AND. But, maybe it does not make a lot of sense to keep it.

pchaigno replied:
No no, I just didn't really think about what I was doing :smile:

kkourt reviewed (commented):
Just some minor comments from a first cursory pass :)

pchaigno commented on bpf/lib/policy.h:
```diff
@@ -320,8 +320,8 @@ static __always_inline int policy_can_egress4(struct __ctx_buff *ctx,
- * Will cause the packet to ignore the policy enforcement layer and
- * be considered accepted despite of the policy outcome.
+ * Will cause the packet to ignore the policy enforcement verdict for allow rules and
+ * be considered accepted despite of the policy outcome. Has no effect on deny rules.
```
That doesn't seem intended. @aanm Is that already on your radar?

aanm replied:
What's not intended?

jrajahalme replied:
@aanm Maybe this updated comment can be revised later if the behavior is changed?

pchaigno replied:
@jrajahalme Yes, definitely not blocking this PR. Thanks for fixing the comment!

pchaigno replied:
I've opened https://github.com/cilium/cilium/issues/18811 for this.

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -85,19 +85,18 @@ encode_custom_prog_meta(struct __ctx_buff *ctx, int ret, __u32 identity)
-					    struct ipv6_ct_tuple *tuple,
-					    int l3_off, struct ipv6hdr *ip6,
-					    __u32 *dst_id)
+static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *dst_id)
```
I'm surprised checkpatch doesn't complain about the line being too long here :thinking:

jrajahalme replied:
I recall it complains of lines longer than 100 characters?

qmonnet replied:
Jarno is correct https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?h=bdc48fa11e46f867ea4d75fa59ee87a7f48be144

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -455,40 +470,12 @@ static __always_inline int ipv6_l3_from_lxc(struct __ctx_buff *ctx,
 {
 	__u32 dst_id = 0;
-	int ret = handle_ipv6(ctx, &dst_id);
+	int ret = handle_ipv6_from_lxc(ctx, &dst_id);
```
Nice! That has been annoying me for so long...

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -110,22 +110,6 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
 	if (!revalidate_data(ctx, &data, &data_end, &ip6))
 		return DROP_INVALID;
 
-	/* Handle special ICMPv6 messages. This includes echo requests to the
```
Commit `datapath: Reduce from LXC complexity` says:

> IPv6 code path is refactored to pattern it more similar to IPv4 code path.

That's not the case anymore, right? The refactoring was moved to other commits?

jrajahalme replied:
Thanks, removed this sentence from the commit message :-)

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -194,6 +143,21 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
+
+		ct_state_new.rev_nat_index = meta >> 16;
+		ct_state_new.loopback = meta & 1;
+		hairpin_flow = ct_state_new.loopback;
```
The logic to encode/decode the values from slot `CB_CT_STATE` could be moved to helper functions. Especially given we have the exact same logic on the IPv4 side.

jrajahalme replied:
Done!

pchaigno reviewed (approved):
Thanks! Only nits below :point_down: 

(Oh, and thanks a lot for the clean commit history. Super easy to review! :pray:)

kkourt reviewed (approved):
LGTM, thanks!

pchaigno commented on bpf/lib/common.h:
```diff
@@ -587,6 +589,7 @@ enum {
 #define	CB_IPCACHE_SRC_LABEL	CB_IFINDEX	/* Alias, non-overlapping */
 	CB_POLICY,
 #define	CB_ADDR_V6_2		CB_POLICY	/* Alias, non-overlapping */
+#define	CB_BACKEND_ID		CB_POLICY       /* Alias, non-overlapping */
```
Nit: This is using spaces instead of tabs between CB_POLICY and the comment.

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -535,12 +535,12 @@ int tail_handle_ipv6(struct __ctx_buff *ctx)
 				return ret;
-
-			/* Store state to be picked up on the continuation tail call. */
-			lb_ctx_store_state(ctx, &ct_state_new);
```
Shouldn't commit `datapath: Be more robust about state between tail calls` be merged with the previous commit? As far as I can see, it only touches lines introduces by the previous commit.

jrajahalme replied:
Good point, squashed :-)

jrajahalme committed:
datapath: Define ENABLE_PER_PACKET_LB as 1 to fix is_defined()

Define ENABLE_PER_PACKET_LB as 1 instead of an empty macro so that
is_defined() works as intended.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme commented on bpf/bpf_lxc.c:
```diff
@@ -52,7 +52,7 @@
 
 #if !defined(ENABLE_HOST_SERVICES_FULL) || defined(ENABLE_SOCKET_LB_HOST_ONLY)
-# define ENABLE_PER_PACKET_LB
+# define ENABLE_PER_PACKET_LB 1
```
This change was needed for `declare_tailcall_if(is_defined(ENABLE_PER_PACKET_LB), ...` to work as expected. Without this change the tailcall got never declared and the declared function was called inline, causing a verifier failure due to too large program.

ti-mo replied:
Hmmm, that sounds like a red herring. Tail calls are never directly invoked and thus cannot be inlined. Are you sure?

pchaigno replied:
No no, that makes sense. `is_defined` expects a value for the macros. We've had this issue in the past.
Then if it returns false, `declare_tailcall_if` wouldn't compile into a tail call.

> Tail calls are never directly invoked and thus cannot be inlined.

Not sure what you mean by this.

ti-mo replied:
Re: inlining: I see, inlining vs. emitting functions as tail calls seems to be the whole point of `declare_tailcall_...`. :slightly_smiling_face: Had never explored the declaration.

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -149,52 +138,42 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
+	/* Determine the destination category for policy fallback.  Service
+	 * translation of the destination address is done before this function,
+	 * so we can do this first. Also, verifier on kernel 4.9 insisted this
+	 * be done before the CT lookup below.
```
I'm guessing the reason we were making that lookup after the CT lookup is to avoid one lookup in case we need to send the packet to the proxy. Why did the verifier want this here? Was it a complexity issue or some other error?

jrajahalme replied:
Complexity issue in kernel 4.9.

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -149,52 +138,42 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
-skip_service_lookup:
+#ifdef ENABLE_PER_PACKET_LB
+#if !defined(DEBUG) && defined(TUNNEL_MODE)
+	/* verifier workaround on kernel 4.9, not needed otherwise */
```
What was the verifier error here? Was it a complexity issue?

jrajahalme replied:
Yes, complexity issue in the test-verifier on kernel 4.9 after the complexity issue on the end-to-end test was fixed by reshuffling the code.

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -347,8 +305,6 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
 	if (!revalidate_data(ctx, &data, &data_end, &ip6))
 		return DROP_INVALID;
 
-	daddr = (union v6addr *)&ip6->daddr;
```
What's the reason for this change?

jrajahalme replied:
Part of limiting the scope of variables in the stack to try resolve the complexity issue on kernel 4.9.

pchaigno replied:
That shouldn't affect complexity. It shouldn't even affect the generated bytecode as far as I can see.

pchaigno reviewed (changes requested):
I'm a bit worried about moving things around because there was a verifier error without really knowing why it complained. Unless it was a complexity issue, there's probably a better fix.

The rest looks good to me.

brb reviewed (approved):
:rocket:

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -120,71 +125,61 @@ ipv6_l3_from_lxc(struct __ctx_buff *ctx, struct ipv6_ct_tuple *tuple,
+#if !defined(DEBUG) && defined(TUNNEL_MODE)
+	/* verifier workaround on kernel 4.9, not needed otherwise */
+	if (!revalidate_data(ctx, &data, &data_end, &ip6))
+		return DROP_INVALID;
```
Ok, so after some testing with this branch and https://github.com/jrajahalme/cilium/pull/new/wip-datapath-reduce-from-lxc-complexity, it turns out the complexity issue isn't "fixed" here because of the helper call, but simply because LLVM reshuffles code around once you add `revalidate_data`. That's why adding `relax_verifier` instead doesn't help.

So this is very fragile and likely to still cause complexity issues elsewhere. I don't have a better solution though. Just note that we don't have complete coverage for complexity issues so we might be adding more here than we are fixing :grimacing:
